library-best-practices/glossary.md

```diff
@@ -66,3 +66,72 @@ Java Dependency Glossary
 - **Linkage-matchable version** (said of a particular version of A in relation
   to all versions of B): There exists some version of B such that the version of
   A and the version of B are linkage-compatible.
+
+
+### Static Linkage Check
+
+- **Static linkage check**: a process to identify the existence of _static
+  linkage conflicts_ for a given classpath, by scanning Java classes and
+  verifying the availability in the classpath.
+
+- **Class usage graph**: a possibly cyclic directed graph of references between classes
+  in the classpath. The nodes (vertices) of the graph correspond to
+  Java classes and the directed edges are references between classes.
+  For example, when 'Class A' has reference to 'Class B' on calling a method X,
+  the class usage graph holds an edge between two nodes:
+
+  ```
+  [Class A] --(method X of class B)-> [Class B]
+  ```
+
+  In this case, 'Class A' is called the _source class_ of the reference and
+  'Class B' is called the _destination class_.
+
+  In general, there can be multiple (parallel) edges between two nodes when
+  a class is calling two or more methods and fields on another class.
+  Self-loops (references between the same class) are possible and
+  common, however, it is safe to omit such references during static linkage checks,
+  because the self-loop references are ensured to be linkage compatible by Java compiler.
+
+- **A reference**: in the class usage graph, the relationship between two 
+  classes is either _class reference_, _method reference_ or _field reference_.
+
+  A _method reference_ indicates that the source class invokes a (static or
+  non-static) method of the destination class.
+
+  A _field reference_ indicates that the source class accesses a (static or
+  non-static) field of the destination class.
+
+  A _class reference_ indicates that the source class uses the destination
+  class without referencing a specific field or method (e.g., class inheritance).
+
+- **Static linkage error**: a reported error for a reference as the result of
+  a static linkage check.
+  A static linkage error contains the information on the source class and
+  the destination class of the reference. Each static linkage error has
+  one of the three types:
+
+  - _Missing class type_ is for errors when where the destination class of a
+    class reference does not exist in the classpath. This error
+    happens when a class is removed in a different version of a library,
+    or there is a dependency missed when constructing the classpath.
+    The reference that causes a missing class error is called a _dangling reference_.
+
+  - _Missing method type_ is for errors when a method reference has a static
+    linkage conflict.
+
+  - _Missing field type_ is for errors when a field reference has a static
+     linkage conflict.
+
+- **Reachability** is the attribute of static linkage errors to indicate
```

Comment:
I think we are tying our terms too much to the "static" type of linkage errors; for several  of these terms, there is no inherent reason that they should be limited to static linkage errors - they could apply to dynamic linkage errors too. 

Also, I feel like "reachability" should be defined as an attribute of references, not linkage errors.

Reply:
agreed